Скачиваем библиотеки

In [1]:
!pip install lxml
!pip install xgboost
!pip install lightgbm
!pip install phik
!pip install pandas
!pip install numpy
!pip install seaborn
!pip install matplotlib
!pip install sklearn
!pip install phik
!pip install random
!pip install -U scikit-learn
!pip install tensorflow


[notice] A new release of pip is available: 23.3.1 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip is available: 23.3.1 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip is available: 23.3.1 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip is available: 23.3.1 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip is available: 23.3.1 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip is available: 23.3.1 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip is available: 23.3.1 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip is available: 23.3.1 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip is available: 23.3.1 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip is available: 23.3.1 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip
ERROR: Could not find a version that satisfies the requirement random (from versions: none)
ERROR: No matching distribution found for random

[notice] A new release of pip is available: 23.3.1 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip


   ---------------------------------------- 0.0/10.6 MB ? eta -:--:--
   ---------------------------------------- 0.0/10.6 MB 660.6 kB/s eta 0:00:17
    --------------------------------------- 0.2/10.6 MB 2.8 MB/s eta 0:00:04
   -- ------------------------------------- 0.5/10.6 MB 4.2 MB/s eta 0:00:03
   --- ------------------------------------ 0.9/10.6 MB 5.2 MB/s eta 0:00:02
   ---- ----------------------------------- 1.3/10.6 MB 5.9 MB/s eta 0:00:02
   ------ --------------------------------- 1.8/10.6 MB 6.6 MB/s eta 0:00:02
   -------- ------------------------------- 2.3/10.6 MB 7.2 MB/s eta 0:00:02
   ---------- ----------------------------- 2.8/10.6 MB 7.8 MB/s eta 0:00:01
   ------------ --------------------------- 3.4/10.6 MB 8.3 MB/s eta 0:00:01
   -------------- ------------------------- 3.7/10.6 MB 8.5 MB/s eta 0:00:01
   ---------------- ----------------------- 4.5/10.6 MB 8.9 MB/s eta 0:00:01
   ----------------- ---------------------- 4.7/10.6 MB 8.8 MB/s eta 0:00:01
   -


[notice] A new release of pip is available: 23.3.1 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip is available: 23.3.1 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip


Импортируем Библиотеки

In [2]:
import requests
import lxml
from bs4 import BeautifulSoup
import pandas as pd
import numpy as mp
import json
from tqdm.notebook import tqdm
import folium
import seaborn as sns
import matplotlib.pyplot as plt
import missingno as msno
import numpy as np
from sklearn.cluster import KMeans, DBSCAN
from sklearn.mixture import GaussianMixture
from sklearn.preprocessing import LabelEncoder, StandardScaler

from phik.report import plot_correlation_matrix
# from sklearn.pipeline import Pipeline

from lightgbm import LGBMClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from random import randint
from sklearn.model_selection import train_test_split, KFold, cross_val_score, RandomizedSearchCV, GridSearchCV
from sklearn.metrics import roc_auc_score, accuracy_score
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline

In [3]:
data = pd.read_csv('data.csv')

In [4]:
data

№п/п        №документа дата регистрации         исходный номер и дата  \
0      11  WSR/1.4-828/2019       08.04.2019              140-р 08.04.2019   
1      22    WSR/1-817/2019       08.04.2019                б/н 05.04.2019   
2      26    WSR/1-813/2019       08.04.2019          ИП-398/05 05.04.2019   
3     114    WSR/1-724/2019       01.04.2019        Д10/4932-ис 07.03.2019   
4     116    WSR/1-722/2019       01.04.2019               3365 25.03.2019   
..    ...               ...              ...                           ...   
303  4936  WSR/1.-1775/2017       03.11.2017             107/10 27.10.2017   
304  4937  WSR/1.-1774/2017       07.11.2017          2040/01/2 17.10.2017   
305  4952  WSR/1.-1761/2017       02.11.2017  03-15-1514/17-0-2 20.10.2017   
306  4953  WSR/1.-1758/2017       02.11.2017     4598/01-021-17 01.11.2017   
307  4961  WSR/1.-1752/2017       01.11.2017      ДО-7548-02-09 31.10.2017   

              Адресат                                              Автор  \
0    Крайчинская С.Б.  Ажгиревич А.И. (ОООР "Союз Машиностроителей Ро...   
1         Уразов Р.Н.  Без автора (Управление Президента Российской Ф...   
2         Уразов Р.Н.  Потехина И.П. (Министерство просвещения Россий...   
3         Уразов Р.Н.  Пашков К.А. (Министерство транспорта Российско...   
4         Глушко Д.Е.  Томских А.А. (Министерство образования, науки ...   
..                ...                                                ...   
303       Уразов Р.Н.                    Климов Е.С. (ООО "Эттон-Центр")   
304       Уразов Р.Н.  Равич К. (Министерство промышленной политики Р...   
305       Уразов Р.Н.  Соляников Ю.В. (Комитет по образованию Правите...   
306       Уразов Р.Н.  Зубов Д.Ю. (Министерство образования и науки К...   
307       Уразов Р.Н.  Запруднова Е.В. (Администрация Владимирской об...   

                                     Краткое содежание  \
0    Вх - Об участии в заочном заседании Совета по ...   
1    Вх - Об участии в заседании рабочей группы по ...   
2    Вх - О предоставлении документов для заключени...   
3    Вх - Об участии в совещании по вопросу достиже...   
4    Вх - О направлении дорожной карты внедрения ре...   
..                                                 ...   
303      Возражения на Отзыв Союза на Исковое завление   
304  О приглашении на IV Чемпионат WorldSkills Hi-T...   
305  Об участии в мероприятиях IV Чемпионата WorldS...   
306  Об участии в IV Национальном чемпионате сквозн...   
307  О присутствии школьников на Мировом Чемпионате...   

                                            Примечание  
0    Повестка заседания, а также материалы к нему, ...  
1    поступило с эл. адреса prm_Bilenkina@gov.ru в ...  
2    срок до 6 апреля, письмо прислали 5 апр в 19:4...  
3          письмо получено 01.04.2019 - (Почта России)  
4                      для передачи Уфимцеву/Дмитриеву  
..                                                 ...  
303  Без приложений, т.к. письмо направлено в Арбит...  
304                            Направлено Мироновой С.  
305                            Направлено Мироновой С.  
306                             Направлено Глушко Д.Е.  
307    Передано Митькиной А., т.к. ответ на ее письмо.  

[308 rows x 8 columns]

# Обучение

Перекодируем данные

In [5]:
# Подключаем класс для предобработки данных
from sklearn import preprocessing

# Напишем функцию, которая принимает на вход DataFrame, кодирует числовыми значениями категориальные признаки
# и возвращает обновленный DataFrame и сами кодировщики.
def number_encode_features(init_df):
    result = init_df.copy() # копируем нашу исходную таблицу
    encoders = {}
    for column in result.columns:
        if result.dtypes[column] == object: # np.object -- строковый тип / если тип столбца - строка, то нужно его закодировать
            encoders[column] = preprocessing.LabelEncoder() # для колонки column создаем кодировщик
            result[column] = encoders[column].fit_transform(result[column]) # применяем кодировщик к столбцу и перезаписываем столбец
    return result, encoders

encoded_data, encoders = number_encode_features(data) # Теперь encoded data содержит закодированные кат. признаки
encoded_data.head()

№п/п  №документа  дата регистрации  исходный номер и дата  Адресат  Автор  \
0    11         268                37                    104       16      4   
1    22         213                37                    276       38     22   
2    26         212                37                    247       38    164   
3   114         204                 1                    236       38    156   
4   116         202                 1                    165        3    200   

   Краткое содежание  Примечание  
0                186         138  
1                199         216  
2                120         239  
3                205         212  
4                 59         198

Задаем в переменные x и у свои данные

In [6]:
X = encoded_data.drop('Адресат', axis=1)
y = encoded_data['Адресат']

Разделим данные на тренировочный и тестовый наборы

In [7]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Создание и обучение модели RandomForest

In [8]:
model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

Предсказание классов для тестового набора данных

In [9]:
y_pred = model.predict(X_test)

Оценка точности модели

In [10]:
accuracy = accuracy_score(y_test, y_pred)
print(f'Точность модели: {accuracy}')

Точность модели: 0.6612903225806451


# Feature Engineering

In [17]:
import pandas as pd
from sklearn.datasets import load_iris
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline



# Создание конвейера с шагами нормализации и классификации
pipeline = Pipeline([
    ('scaler', StandardScaler()),  # Нормализация данных
    ('rf', model)  # Модель RandomForest
])

# Определение параметров для Grid Search
param_grid = {
    'rf__n_estimators': [10, 50, 100],
    'rf__max_depth': [ 10, 20, 30],
    'rf__min_samples_split': [2, 5, 10],
    'rf__min_samples_leaf': [1, 2, 4]
}

# Создание объекта Grid Search
grid_search = GridSearchCV(pipeline, param_grid=param_grid, cv=5, scoring='accuracy')

# Обучение модели с использованием Grid Search
grid_search.fit(X_train, y_train)

# Вывод оптимальных параметров
print("Оптимальные параметры:", grid_search.best_params_)

# Оценка производительности модели на тестовых данных
accuracy = grid_search.score(X_test, y_test)
print(f"Точность на тестовых данных: {accuracy:.2f}")


c:\Users\tooer\lwjgl\.venv\Lib\site-packages\sklearn\model_selection\_split.py:737: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(


Оптимальные параметры: {'rf__max_depth': 10, 'rf__min_samples_leaf': 1, 'rf__min_samples_split': 2, 'rf__n_estimators': 100}
Точность на тестовых данных: 0.66


# Выводы

В результате этой сессии было выполнено обучение модули с оценкой точности 0.6612903225806451 это больше по сравнению с другими моделями приблизительно на 0.15.После Feature Engineering точность модели стала 0.6451612903225806